File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/MIDI/MIDIView.swift
import SwiftUI
import VoxaTrace

/// MIDI synthesis view demonstrating SonixMidiSynthesizer.Builder pattern.
struct MIDIView: View {
    @StateObject private var viewModel = MIDIViewModel()

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("MIDI Synthesis")
                .font(.headline)

            Text("Status: \(viewModel.status)")
                .font(.caption)
                .foregroundColor(.secondary)

            // Show FluidSynth version if available
            if let version = viewModel.synthesizerVersion {
                Text("FluidSynth: \(version)")
                    .font(.caption)
                    .foregroundColor(.secondary)
            }

            // C major scale notes
            Text("Notes: C4 - D4 - E4 - F4 - G4 - A4 - B4 - C5")
                .font(.caption)
                .foregroundColor(.secondary)

            // Sample rate selection
            sampleRateSelector

            // Synthesize button
            Button(viewModel.isSynthesizing ? "Synthesizing..." : "Synthesize") {
                viewModel.synthesize()
            }
            .disabled(viewModel.isSynthesizing)
            .buttonStyle(.borderedProminent)

            // Playback controls
            playbackControls
        }
        .onDisappear {
            viewModel.onDisappear()
        }
    }

    private var sampleRateSelector: some View {
        HStack {
            Text("Sample Rate:")
                .font(.caption)
            ForEach(viewModel.sampleRates, id: \.self) { rate in
                sampleRateButton(rate: rate)
            }
        }
    }

    @ViewBuilder
    private func sampleRateButton(rate: Int32) -> some View {
        let label = "\(rate / 1000)kHz"
        let isSelected = viewModel.selectedSampleRate == rate
        if isSelected {
            Button(label) {
                viewModel.selectedSampleRate = rate
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.small)
            .disabled(viewModel.isSynthesizing)
        } else {
            Button(label) {
                viewModel.selectedSampleRate = rate
            }
            .buttonStyle(.bordered)
            .controlSize(.small)
            .disabled(viewModel.isSynthesizing)
        }
    }

    private var playbackControls: some View {
        HStack(spacing: 12) {
            Button("Play") {
                viewModel.play()
            }
            .disabled(!viewModel.canPlay)
            .buttonStyle(.bordered)

            Button("Stop") {
                viewModel.stop()
            }
            .disabled(!viewModel.isPlaying)
            .buttonStyle(.bordered)
        }
    }
}

/// Backward compatibility alias.
typealias MidiSection = MIDIView

#Preview {
    MIDIView()
        .padding()
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/midi/view/MIDIView.kt
package com.musicmuni.voxatrace.demo.sections.midi.view

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.musicmuni.voxatrace.demo.components.OptionChip
import com.musicmuni.voxatrace.demo.sections.midi.viewmodel.MIDIViewModel

/**
 * MIDI View - UI for MIDI synthesis demo.
 *
 * Demonstrates:
 * - SonixMidiSynthesizer Builder pattern
 * - Custom sample rate configuration
 * - synthesizeFromNotes() for programmatic MIDI generation
 * - SonixPlayer for playback of synthesized audio
 */
@Composable
fun MIDIView(viewModel: MIDIViewModel = viewModel()) {
    val context = LocalContext.current

    // Stop playback when leaving the screen (matches iOS onDisappear)
    DisposableEffect(Unit) {
        onDispose {
            viewModel.stop()
        }
    }

    // Collect state from ViewModel
    val status by viewModel.status.collectAsStateWithLifecycle()
    val isSynthesizing by viewModel.isSynthesizing.collectAsStateWithLifecycle()
    val outputFile by viewModel.outputFile.collectAsStateWithLifecycle()
    val synthesizerVersion by viewModel.synthesizerVersion.collectAsStateWithLifecycle()
    val selectedSampleRate by viewModel.selectedSampleRate.collectAsStateWithLifecycle()
    val isPlaying by viewModel.isPlaying.collectAsStateWithLifecycle()

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "MIDI Synthesis",
            style = MaterialTheme.typography.titleMedium
        )

        Text(text = "Status: $status", style = MaterialTheme.typography.bodySmall)

        // Show FluidSynth version if available
        synthesizerVersion?.let {
            Text(text = "FluidSynth: $it", style = MaterialTheme.typography.bodySmall)
        }

        // C major scale notes
        Text(
            text = "Notes: C4 - D4 - E4 - F4 - G4 - A4 - B4 - C5",
            style = MaterialTheme.typography.bodySmall
        )

        // Sample rate selection
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Sample Rate:", style = MaterialTheme.typography.bodySmall)
            listOf(44100, 48000).forEach { rate ->
                OptionChip(
                    selected = selectedSampleRate == rate,
                    onClick = { viewModel.setSelectedSampleRate(rate) },
                    label = "${rate / 1000}kHz",
                    enabled = !isSynthesizing
                )
            }
        }

        // Synthesize button
        Button(
            onClick = { viewModel.synthesize(context) },
            enabled = !isSynthesizing
        ) {
            Text(if (isSynthesizing) "Synthesizing..." else "Synthesize")
        }

        // Playback controls
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = { viewModel.play() },
                enabled = outputFile != null && !isSynthesizing && !isPlaying
            ) {
                Text("Play")
            }

            Button(
                onClick = { viewModel.stop() },
                enabled = isPlaying
            ) {
                Text("Stop")
            }
        }
    }
}
